ingest-check-preservica-for-existing-io/src/main/scala/uk/gov/nationalarchives/Lambda.scala

```diff
@@ -0,0 +1,84 @@
+package uk.gov.nationalarchives
+
+import cats.effect.*
+import io.circe.generic.auto.*
+import pureconfig.ConfigReader
+import pureconfig.generic.derivation.default.*
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue
+import sttp.capabilities.fs2.Fs2Streams
+import uk.gov.nationalarchives.DADynamoDBClient.DADynamoDbRequest
+import uk.gov.nationalarchives.DynamoFormatters.*
+import uk.gov.nationalarchives.DynamoFormatters.Type.*
+import uk.gov.nationalarchives.Lambda.*
+import uk.gov.nationalarchives.dp.client.EntityClient
+import uk.gov.nationalarchives.dp.client.EntityClient.*
+import uk.gov.nationalarchives.dp.client.EntityClient.EntityType.*
+import uk.gov.nationalarchives.dp.client.fs2.Fs2Client
+
+import java.util.UUID
+
+class Lambda extends LambdaRunner[Input, StateOutput, Config, Dependencies] {
+  private val sourceId = "SourceID"
+  override def handler: (
+      Input,
+      Config,
+      Dependencies
+  ) => IO[StateOutput] = (input, config, dependencies) =>
+    for {
+      assetItems <- dependencies.dynamoDbClient.getItems[AssetDynamoTable, PartitionKey](
+        List(PartitionKey(input.id)),
+        config.dynamoTableName
+      )
+      asset <- IO.fromOption(assetItems.headOption)(
+        new Exception(s"No asset found for ${input.id} from ${input.batchId}")
+      )
+      _ <- IO.raiseWhen(asset.`type` != Asset)(
+        new Exception(s"Object ${asset.id} is of type ${asset.`type`} and not 'Asset'")
+      )
+      logCtx = Map("batchId" -> input.batchId, "assetId" -> asset.id.toString)
+      log = logger.info(logCtx)(_)
+      _ <- log(s"Asset ${asset.id} retrieved from Dynamo")
+
+      entitiesWithAssetName <- dependencies.entityClient.entitiesByIdentifier(Identifier(sourceId, asset.name))
+
+      potentialIoExistsResponse = entitiesWithAssetName.headOption match {
+        case Some(entityWithAssetName) =>
+          val potentialEntityType: Option[EntityType] = entityWithAssetName.entityType
+          if (potentialEntityType.contains(InformationObject)) Right(true)
+          else
+            Left(
+              new Exception(
```

Comment:
Do you need to throw an exception if it's not an InformationObject? Is it possible that anything but an IO will have a SourceId attached or is it possible in the future that we might legitimately want to put a SourceId on another entityType? I'm not sure but it feels like the purpose of this lambda is to check if there is an IO with that source ID. If it's not an IO, it returns false. 

The other thing is that we do sometimes get odd results from Preservica so being as permissive as possible is usually good.

Also, we've got tests on the opex creator which should check that we're putting the SourceId in the right place. 

What do you think? You could simplify this a fair bit as well because it would just become:

```scala
assetExists = entitiesWithAssetName.headOption.contains(InformationObject)
```

Reply:
Ok. yeah, this was a last minute addition. I was thinking thinking in case we accidentally/purposely add a `sourceId` to a folder or file. I'll change it back